src/Nethermind/Nethermind.Optimism.Test/ReceiptDecoderTests.cs

```diff
@@ -0,0 +1,72 @@
+// SPDX-FileCopyrightText: 2022 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System.Collections;
+using Nethermind.Core.Extensions;
+using Nethermind.Serialization.Rlp;
+using NUnit.Framework;
+
+namespace Nethermind.Optimism.Test;
+
+public partial class ReceiptDecoderTests
+{
+    [TestCaseSource(nameof(DepositTxReceiptsSerializationTestCases))]
+    public void Can_build_correct_block_tree(byte[] rlp, bool includesNonce, bool includesVersion, bool shouldIncludeNonceAndVersionForTxTrie)
+    {
+        OptimismReceiptMessageDecoder decoder = new();
+        OptimismTxReceipt decodedReceipt = decoder.Decode(new RlpStream(rlp), RlpBehaviors.SkipTypedWrapping);
+        Assert.That(decodedReceipt.DepositNonce, includesNonce ? Is.Not.Null : Is.Null);
+        Assert.That(decodedReceipt.DepositReceiptVersion, includesVersion ? Is.Not.Null : Is.Null);
+
+        RlpStream encodedRlp = new(decoder.GetLength(decodedReceipt, RlpBehaviors.SkipTypedWrapping));
+        decoder.Encode(encodedRlp, decodedReceipt, RlpBehaviors.SkipTypedWrapping);
+
+        Assert.That(rlp, Is.EqualTo(encodedRlp.Data.ToArray()));
+
+        OptimismReceiptTrieDecoder trieDecoder = new();
+        RlpStream encodedTrieRlp = new(trieDecoder.GetLength(decodedReceipt, RlpBehaviors.SkipTypedWrapping));
+
+        trieDecoder.Encode(encodedTrieRlp, decodedReceipt, RlpBehaviors.SkipTypedWrapping);
+        encodedTrieRlp.Position = 0;
+
+        OptimismTxReceipt decodedTrieReceipt = trieDecoder.Decode(encodedTrieRlp, RlpBehaviors.SkipTypedWrapping);
+
+        Assert.That(decodedTrieReceipt.DepositNonce, shouldIncludeNonceAndVersionForTxTrie ? Is.Not.Null : Is.Null);
+        Assert.That(decodedTrieReceipt.DepositReceiptVersion, shouldIncludeNonceAndVersionForTxTrie ? Is.Not.Null : Is.Null);
```

Comment:
I see 3 different assert points, I understand it is easier to keep it as one test, but maybe break it down to 3 local method calls?

Reply:
Made two local ones, hope looks OK